feat(lua): expose lua-cjson as vim.json

mjlbach opened the pull request:
This adds a direct to/from lua json encode/decode using lua-cjson. There are performance benefits to having direct conversion from json to the lua datastructures, and this avoids any issue calling in luv callbacks.

- [x] replace cjson.null with vim.NIL
- [x] add to lua interpreter
- [x] figure out where to put license/if having the license info in the file header s fine
- [x] address fpconv warnings (silence)
- [x] address justin's concern with circular import
- [x] add real benchmark
- [x] add more tests

lithammer commented:
I'm curious what the rationale for this is? What does it offer that `json_encode`/`json_decode` doesn't?

hrsh7th commented:
In my knowledge, vim.fn.json_encode/json_decode needs `vim.schedule` and the return value needs conversion between vim and lua.

mjlbach commented:
@hrsh7th is correct, I'd also like to add in my benchmark lua-cjson was able to decode the same file in ~%30 of the time.

Sh3Rm4n commented:
And what is the rationale to embed / vendor the lua_cjson code? Is it just for testing? Is it similar to treesitter which was first vendored, and after it matured a bit outsourced again to be able to use the upstream version?

bfredl commented:
To make it work directly with our lua mpack typesystem, i e `vim.NIL` and `vim.empty_dict()`

justinmk commented on src/nvim/json/lua_cjson.c:
```diff
@@ -0,0 +1,1590 @@
+
+/* Caveats:
+ * - JSON "null" values are represented as lightuserdata since Lua
+ *   tables cannot contain "nil". Compare with cjson.null.
```
we removed use of lightuserdata in https://github.com/neovim/neovim/pull/7880

mjlbach replied:
Definitely going to remove the internal cjson null representation and replace with vim.NIL, so no lightuserdata when done (that's why we have to vendor this instead of just adding the library), just haven't finished yet :)

bfredl commented:
I'm looking into handling NULL and empty tables correctly (according to our conventions).

bfredl commented:
Added to interpreter to `vim.json` (could be made lazy, later). Initial support for `vim.NIL` and `vim.empty_dict` is prototyped.

tami5 commented:
😍 boy this is exciting

justinmk commented:
How feasible is it to remove the old Vimscript implemenation of `json_encode()` / `json_decode()` and make it a wrapper around this. Even if it makes Vimscript slightly slower, the code/maintenance reduction is attractive...

mjlbach commented:
I'll pick this PR back up this weekend.

It's feasible but we should also consider benchmarking it to see if the conversion cost to vim data structures is high.

justinmk commented on src/nvim/json/fpconv.c:
```diff
@@ -0,0 +1,211 @@
+/* fpconv - Floating point conversion routines
```
Similar to [src/xdiff/](https://github.com/neovim/neovim/issues/11239) and [src/mpack/](https://github.com/neovim/neovim/issues/15566),  it would make sense for this subsystem to live in `src/json/` . The rationale is that this signals to contributors that we don't own these modules. We may add a few tweaks such as `mpack.NIL` handling, but otherwise try to keep this code sync'd with upstream. And try to avoid dependencies on `src/nvim/` (if possible).

see also
- https://github.com/neovim/neovim/issues/11239 
- https://github.com/neovim/neovim/pull/15566#discussion_r703069659

clason replied:
Added benefit: make it easier/clearer to apply Neovim style guide (uncrustify) only `src/nvim`, not external subsystems.

mjlbach replied:
done

mjlbach commented:
I rebased, cleaned up some conflicts, and moved the source of cjson as requested. I'll squash the commits when everything is ready. I guess tests are left, and silencing some of the missing/strict prototype warnings.

mjlbach commented:
I silenced the warnings (not sure if this was the cleanest way) and added some basic tests.

justinmk commented on src/cjson/lua_cjson.h:
```diff
@@ -0,0 +1,8 @@
+#ifndef NVIM_LUA_CJSON_H
+#define NVIM_LUA_CJSON_H
```
the header guards should be renamed, `NVIM_` prefix doesn't match their path.

mjlbach replied:
Can you clarify what you would like the guard to be? `CJSON_H`?

justinmk replied:
`CJSON_LUACJSON_H` i guess

mjlbach replied:
Ah, got it. Done.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -186,6 +186,10 @@ if(NOT MSVC)
+  # lua-cjson: ignore strict and missing prototypes for cjson
+  file(GLOB CJSON_SOURCES ../cjson/*.c)
+  set_source_files_properties(
+    ${CJSON_SOURCES} PROPERTIES COMPILE_FLAGS "${COMPILE_FLAGS} -Wno-strict-prototypes -Wno-missing-prototypes -Wno-conversion")
```
"missing prototypes" means there is a build ordering issue or the headers are missing. shouldn't ignore it

mjlbach replied:
I fixed the missing prototypes warning, (it was from unused functions), are we ok with silencing strict prototypes?

mjlbach replied:
For context, the remaining warnings are from strict prototypes in the floating point conversion files, which I did not modify: 
```
[311/334] Building C object src/nvim/CMakeFiles/nvim.dir/__/cjson/fpconv.c.o
In file included from ../src/cjson/fpconv.c:36:
../src/cjson/fpconv.h:15:1: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
   15 | extern void fpconv_init();
      | ^~~~~~
../src/cjson/fpconv.c:58:13: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
   58 | static void fpconv_update_locale()
      |             ^~~~~~~~~~~~~~~~~~~~
../src/cjson/fpconv.c:205:6: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
  205 | void fpconv_init()
      |      ^~~~~~~~~~~
[315/334] Building C object src/nvim/CMakeFiles/nvim.dir/__/cjson/lua_cjson.c.o
In file included from ../src/cjson/lua_cjson.c:51:
../src/cjson/fpconv.h:15:1: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
   15 | extern void fpconv_init();
      | ^~~~~~
[334/334] Generating doc/tags
```

kylo252 replied:
For reference, clang is complaining about two different warnings
> clang: No previous prototype for function 'fpconv_init' [-Wmissing-prototypes]
clang: This old-style function definition is not preceded by a prototype [-Wstrict-prototypes]

mjlbach replied:
I know what the warnings are, and I already resolved the former.

justinmk commented on src/cjson/lua_cjson.c:
```diff
@@ -0,0 +1,1609 @@
+#include <lua.h>
+#include <lauxlib.h>
+
+#include "nvim/lua/executor.h"
```
can we avoid this circular dependency? I don't see any `nlua_*` calls below...

mjlbach replied:
The call is here: https://github.com/neovim/neovim/pull/14871/files#diff-770e34bd301f419ac67cceed5626e9e6818d933bf5ba20b453d1530fe1d2b4d3R771

justinmk commented:
before merging can this be squashed such that the upstream code is in 1 commit, and our modifications are in another commit

mjlbach commented:
Yes, once people feel this is in a mergeable state I will clean up the commit history. There will be two commits:

* feat: add lua-cjson as an external dependency
* feat: expose lua-cjson as vim.json (this will include benchmarks, the test, and the modified code)

mjlbach committed:
feat(lua): add lua-cjson as vendored dependency

Derived from the openresty lua-cjson fork at commit https://github.com/openresty/lua-cjson/commit/3d93d297092172eac3d52a1b3b6c1d479da5341a

mjlbach committed:
feat(lua): expose lua-cjson as vim.json

* add vim.json.encode and vim.json.decode
* use vim.NIL instead of cjson.null
* resolve strict-prototypes warnings

* The following benchmark shows an approximately 2.5x (750 ms vs 300 ms) improvement in deserialization performance over
  vim.fn.json_decode on a medium package.json

  ```lua
  local uv = vim.loop
  local function readfile(path)
    return
  end

  local json_url = "https://raw.githubusercontent.com/rust-analyzer/rust-analyzer/b24c8d5c89ee93d1172b4127564f5da3b0c88dad/editors/code/package.json"
  io.popen(string.format('curl -v -f -L -O %q &> /dev/null', json_url))

  local json_string = io.open('package.json'):read '*a'

  uv.update_time()
  local start = uv.hrtime()

  for i = 1,1000 do
    vim.fn.json_decode(json_string)
  end

  uv.update_time()
  print(string.format("Deserialization time vim.fn.json_decode: %s ms", (uv.hrtime() - start) * (1e-6)))

  uv.update_time()
  local start = uv.hrtime()

  for i = 1,1000 do
    vim.json.decode(json_string)
  end
  uv.update_time()
  print(string.format("Deserialization time vim.json.decode: %s ms", (uv.hrtime() - start) * (1e-6)))
  ```

Co-Authored-By: Björn Linse <bjorn.linse@gmail.com>

mjlbach committed:
test: add tests for vim.json

mjlbach commented:
I need to add the real benchmarks, but I cleaned up the commits. There's still the open warnings from fpconv to take care of, and Justin had some concerns about a circular dependency, but I pointed out where the included functions are used.

justinmk commented on test/functional/lua/json_spec.lua:
```diff
@@ -0,0 +1,26 @@
+
+    eq(
+      { hello = 'world'},
+      exec_lua([[return vim.json.decode('{"hello":"world"}')]])
```
i kinda assumed more tests were on the way. Maybe copy a few from `json_functions_spec.lua` ? 

Or parameterize (most of) the tests in `json_functions_spec.lua` so that it tests both.

mjlbach replied:
Looking through the tests, I think parameterizing them is going to be non-trivial, seems like many of them deal with vimscript specific behavior, or eval-ing a string representation of the vim command.

mjlbach commented:
Added the benchmark, added some tests from json functional test. Ready for review and final comments.

clason commented:
Functional test failure:
```
2021-09-26T07:48:44.1162119Z [  ERROR   ] 1 error, listed below:
2021-09-26T07:48:44.1162872Z [  ERROR   ]  @ : test/functional/lua/json_spec.lua
2021-09-26T07:48:44.1164625Z test/functional/lua/json_spec.lua:66: nesting of [[...]] is deprecated near '['
```

clason commented:
@mjlbach I took the liberty of fixing the test with the minimal (whitespace) change. (I did what I had to, because I could!)

Hope you don't mind; feel free to drop the commit and make a different change.

mjlbach commented:
I actually wasn't planning on squashing the test + lua-cjson commits together, but keeping them separate to delineate the c-side changes/modifications.

I just squashed the test commit together so if there aren't any objects I'll merge this after @bfredl or @justinmk approves.

bfredl commented on test/functional/lua/json_spec.lua:
```diff
@@ -0,0 +1,129 @@
+   end)
+
+   it('dumps dictionaries', function()
+     eq('[]', exec_lua([[return vim.json.encode({})]]))
```
` vim.json.encode(vim.empty_dict())`

mjlbach replied:
done, squashed.

bfredl commented on test/functional/lua/json_spec.lua:
```diff
@@ -0,0 +1,129 @@
+     eq('[]', exec_lua([[return vim.json.encode({})]]))
+     eq('{"d":[]}', exec_lua([[return vim.json.encode({d={}})]]))
+   end)
+
```
add a test for `vim.json.encode(vim.NIL)`

mjlbach replied:
done, squashed.

bfredl commented:
otherwise LGTM :)
